### PR TITLE
test/venv: Fix the python version of the venv test environment

### PR DESCRIPTION
**Pull request by wjwithagen**
Otherwise it will result in:
[ 59%] preparing venv for mgr-dashboard
New python executable in /home/jenkins/workspace/ceph-master/build/mgr-dashboard-virtualenv/bin/python2.7
Also creating executable in /home/jenkins/workspace/ceph-master/build/mgr-dashboard-virtualenv/bin/python
Installing setuptools, pip, wheel...done.
Running virtualenv with interpreter /usr/local/bin/python2.7

Even though the Cmake options contained:
  '-DWITH_PYTHON2=OFF' '-DWITH_PYTHON3=3' '-DMGR_PYTHON_VERSION=3'

Mostly due to the fact that
  src/tools/setup-virtualenv.sh
default to:
  PYTHON_BINARY="python2.7"

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
@wjwithagen what's the failure/symptom the change is supposed to address?

**Comment by wjwithagen** [comment]
@tchaikov 
`cmake/modules/AddCephTest.cmake` uses a script:  `src/tools/setup-virtualenv.sh`
to build the virrtual env.
But that script defaults to using python2.7 when installing the python interpreter.
Loic has already forseen other versions, hence the `--python` option.
And using the wrong python version in ceph, on FreeBSD blows up really horribly.

**Comment by wjwithagen** [comment]
@epuertat 
I'm not sure about that this would be the final fix, and I do not have to revert #28696.
Still testing that.
But if I do not revert, and not throw away  `requirements-{py27,p3}.txt` `pip` complains about inconsistent requirement for `pylint`. Which I understand.
I guess that these files stem from the old code, and things really get mixed up.
Bottomline on #28696: got to do some more testing.

**Comment by tchaikov** [comment]
@wjwithagen could you define "And using the wrong python version in ceph, on FreeBSD blows up really horribly." ?

**Comment by epuertat** [comment]
> @wjwithagen could you define "And using the wrong python version in ceph, on FreeBSD blows up really horribly." ?

@tchaikov: @wjwithagen sent an e-mail to ceph dev mail list, with the following Jenkins run: https://github.com/ceph/ceph/pull/29761#issuecomment-523090797

**Comment by epuertat** [comment]
> But if I do not revert, and not throw away `requirements-{py27,p3}.txt` `pip` complains about inconsistent requirement for `pylint`. Which I understand.
> I guess that these files stem from the old code, and things really get mixed up.
> Bottomline on #28696: got to do some more testing.

`requirements-py{27,3}.txt` shouldn't be there, as they've been removed by #28696 (BTW, they were never pre-installed, as `install-deps.sh` and `setup-virtualenv.sh` looked for `*requirements.txt`, but not for `*requirements*.txt`. Now the version locking is set through `constraints.txt` file (following [PEP 508](https://www.python.org/dev/peps/pep-0508/) syntax). So no reason for conflicting `pylint` version in a virtualenv.

I'd like to help on this, but don't have a FreeBSD sandbox to test.

**Comment by epuertat** [comment]
BTW @wjwithagen, dashboard tox test is now stand-alone, and no longer needs any binaries from other Ceph components (librados, librbd, etc.), so you may just run it after the `do_cmake.sh` stuff without any `make` (`ctest -R run-tox-mgr-dashboard`). That should speed up the testing/fix.

**Comment by wjwithagen** [comment]
@epuertat 
Oke I have a succesfull run, with regards to "not reverting #28696:
http://cephdev.digiware.nl:8180/jenkins/job/ceph-master/3661/console

And I know I can run ctests individually, but doing so is cumbersome in my Jenkins env, where I build a patched master to incorporate certain required PRs that a not yet committed to master.
And in the end, the Jenkins run is conclusive about PRs working or not.

Now we'll have to see if @tchaikov likes this patch, or prefers to solve it in a different way.

**Comment by wjwithagen** [comment]
@epuertat 
But the actual dashboard test still fails:
```
            # pylint: disable=broad-except
            except Exception:
                try:
>                   f = open(idp_metadata, 'r')
E                   OSError: [Errno 63] File name too long: '<?xml version="1.0"?>\n<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata"\n                     xmlns:ds="http://www.w3.org/2000/09/xmldsig#"\n                     entityID="https://testidp.ceph.com/simplesamlphp/saml2/idp/metadata.php"\n                     ID="pfx8ca6fbd7-6062-d4a9-7995-0730aeb8114f">\n  <ds:Signature>\n    <ds:SignedInfo>\n      <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>\n      <ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>\n      <ds:Reference URI="#pfx8ca6fbd7-6062-d4a9-7995-0730aeb8114f">\n        <ds:Transforms>\n          <ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>\n          <ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>\n        </ds:Transforms>\n        <ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>\n        <ds:DigestValue>v6V8fooEUeq/LO/59JCfJF69Tw3ohN52OGAY6X3jX8w=</ds:DigestValue>\n      </ds:Reference>\n    </ds:SignedInfo>\n    <ds:SignatureValue>IDP_SIGNATURE_VALUE</ds:SignatureValue>\n    <ds:KeyInfo>\n      <ds:X509Data>\n        <ds:X509Certificate>IDP_X509_CERTIFICATE</ds:X509Certificate>\n      </ds:X509Data>\n    </ds:KeyInfo>\n  </ds:Signature>\n  <md:IDPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">\n    <md:KeyDescriptor use="signing">\n      <ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">\n        <ds:X509Data>\n          <ds:X509Certificate>IDP_X509_CERTIFICATE</ds:X509Certificate>\n        </ds:X509Data>\n      </ds:KeyInfo>\n    </md:KeyDescriptor>\n    <md:KeyDescriptor use="encryption">\n      <ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">\n        <ds:X509Data>\n          <ds:X509Certificate>IDP_X509_CERTIFICATE</ds:X509Certificate>\n        </ds:X509Data>\n      </ds:KeyInfo>\n    </md:KeyDescriptor>\n    <md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"\n                            Location="https://testidp.ceph.com/simplesamlphp/saml2/idp/SingleLogoutService.php"/>\n    <md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:transient</md:NameIDFormat>\n    <md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"\n                            Location="https://testidp.ceph.com/simplesamlphp/saml2/idp/SSOService.php"/>\n  </md:IDPSSODescriptor>\n</md:EntityDescriptor>'
```
so it looks like `idp_metadata` actually does not contain something that we an open. 8-D

**Comment by epuertat** [comment]
@wjwithagen There are 2 issues here:
1. Ceph-Dashboard SSO/SAML2 module includes a Python method that drives the Python lemma "Better ask forgiveness than permission" to its ultimate consequences (too far IMHO). Basically: 
```python
url_or_xml = "<?xml version="1.0"?>..."

try:
  # first let's assume this free-form user-provided data is a URL, and let's do some TCP roundtrips
  urlopen(url_or_xml)
except Exception:
  try:
    # nope? then sure it'll be a local file...
    open(url_or_xml)
  except FileNotFoundError:
    # Bad luck! What a day. It was an XML payload then?
    parse_xml(url_or_xml)
```
2. No matter how weird that code looks, this is not failing in Linux so... The difference lies in the exception that the `open()` is triggering. FreeBSD seems to be very stiff when it comes to filename length, and considers the following as too much for a filename: `"<?xml version="1.0" encoding="UTF-8"?><md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" xmlns:ds="http://www.w3.org/2000/09/xmldsig#" entityID="https://testidp.ceph.com/simplesamlphp/saml2/idp/metadata.php" ID="pfx8ca6fbd7-6062-d4a9-7995-0730aeb8114f"><ds:Signature><ds:SignedInfo><ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" /><ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256" /><ds:Reference URI="#pfx8ca6fbd7-6062-d4a9-7995-0730aeb8114f"><ds:Transforms><ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature" /><ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" /></ds:Transforms><ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256" /><ds:DigestValue>v6V8fooEUeq/LO/59JCfJF69Tw3ohN52OGAY6X3jX8w=</ds:DigestValue></ds:Reference></ds:SignedInfo>ds:SignatureValue>IDP_SIGNATURE_VALUE</ds:SignatureValue><ds:KeyInfo><ds:X509Data><ds:X509Certificate>IDP_X509_CERTIFICATE</ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature><md:IDPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol"><md:KeyDescriptor use="signing"><ds:KeyInfo><ds:X509Data><ds:X509Certificate>IDP_X509_CERTIFICATE</ds:X509Certificate></ds:X509Data>
         </ds:KeyInfo></md:KeyDescriptor>
      <md:KeyDescriptor use="encryption"><ds:KeyInfo><ds:X509Data><ds:X509Certificate>IDP_X509_CERTIFICATE</ds:X509Certificate></ds:X509Data></ds:KeyInfo></md:KeyDescriptor><md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://testidp.ceph.com/simplesamlphp/saml2/idp/SingleLogoutService.php" />
      <md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:transient</md:NameIDFormat><md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://testidp.ceph.com/simplesamlphp/saml2/idp/SSOService.php" /></md:IDPSSODescriptor></md:EntityDescriptor>.xml"`.

So it raises an `OSError` instead of a `FileNotFoundError'.

A question remains: why wasn't this being reported in FreeBSD previously?

As a workaround you may add the following to the `src/pybind/mgr/dashboard/services/sso.py:210`:
```diff
-            except FileNotFoundError:
+            except (FileNotFoundError, OSError):
```

I'll send a fix for this.

**Comment by wjwithagen** [comment]
@epuertat 
Yes, exactly on implementation details there are destinctive differences between FreeBSD en Linux.
Something likewise on error reporting that is not specified in Posix standards..
And more of these sorts.
I'll test your suggested fix

**Comment by wjwithagen** [comment]
@epuertat 
Patch seems to work:

```
(cd build; sudo -u jenkins ctest -R run-tox-mgr-dashboard )
Test project /home/jenkins/workspace/ceph-master/build
    Start 1: run-tox-mgr-dashboard
load: 16.75  cmd: python3.6 36199 [running] 17.97r 12.03u 0.87s 73% 124340k
1/1 Test #1: run-tox-mgr-dashboard ............   Passed  186.81 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 187.31 sec
```

**Comment by wjwithagen** [comment]
> 1. No matter how weird that code looks, this is not failing in Linux so... 
> The difference lies in the exception that the `open()` is triggering. 
> FreeBSD seems to be very stiff when it comes to filename length, and considers the following as too much for a filename: `"<?xml version="1.0 ......................." `

Max path length on FreeBSD is limitted to 1K bytes.
`syslimits.h:#define     PATH_MAX                 1024   /* max bytes in pathname */`

**Comment by wjwithagen** [comment]
> 
> 
> @wjwithagen could you define "And using the wrong python version in ceph, on FreeBSD blows up really horribly." ?

Turns out that "horribly" was not quite caused by this problem, but more the result of the way the test works.... Dumping a lot of XML in 'open()' made me really wonder what was going on....

One the less this PR will be required on systems where python 2.7 and 3.x co-exist. Otherwise the virtual env tests will always be run with python-2.7

File: cmake/modules/AddCephTest.cmake [hunk 1]
```diff
@@ -74,7 +74,7 @@ function(add_tox_test name)
   string(REPLACE ";" "," tox_envs "${tox_envs}")
   add_custom_command(
     OUTPUT ${venv_path}/bin/activate
-    COMMAND ${CMAKE_SOURCE_DIR}/src/tools/setup-virtualenv.sh ${venv_path}
+    COMMAND ${CMAKE_SOURCE_DIR}/src/tools/setup-virtualenv.sh --python=python${Python_VERSION_MAJOR}.${Python_VERSION_MINOR} ${venv_path}
```
Comment (tchaikov) [inline cmake/modules/AddCephTest.cmake]: i agree that we cannot assume the existence of python2.7 on the machine where `make check` runs.

but can you help me understand why we are using `python${Python_VERSION_MAJOR}.${Python_VERSION_MINOR}` here?

as `Python_VERSION_MAJOR` and `Python_VERSION_MINOR` are set by the latest `find_package(Python ...)` call. and `add_tox_test()` could be called anywhere in the source tree on demand. 

can we reply on the fact that `find_package(Python ...)` is called before `add_tox_test()` ? and the venv used by tox are used solely for bootstrap tox env, so any python version supporting tox would work. so i think we might want to just put
```cmake
find_package(Python REQUIRED)
add_custom_command(
    OUTPUT ${venv_path}/bin/activate
    COMMAND ${CMAKE_SOURCE_DIR}/src/tools/setup-virtualenv.sh --python="${Python_EXECUTABLE}" "${venv_path}"
    WORKING_DIRECTORY ${tox_path}
    COMMENT "preparing venv for ${name}")
```

also, the commit message does not make sense. please note, the help messages of these options read:
- `WITH_PYTHON2` "build python2 bindings"
- `WITH_PYTHON3` "build python2 bindings"

i don't think any of them is connected to the python version we used for preparing the venv for tox for testing. not to mention `MGR_PYTHON_VERSION`.

Comment (wjwithagen) [reply cmake/modules/AddCephTest.cmake]: @tchaikov
I saw that `find_package(Python ...)` is called rather late in the whole configuration. It has given me some doubts already on previous instances.
Perhaps more sensible to move this forward, since with python nothing works in the end.
And put it in `./CMakeLists.txt`

For installing the tox-env with 2.7 when actually the build is for 3.6....
When that happens, I think I saw 2.7 modules being pulled in.

Bottom line is that this patch just makes sure that things are done as asked by the user.

So how about moving the `find_package(Python ...)` stuff up the tree?

Comment (tchaikov) [reply cmake/modules/AddCephTest.cmake]: > Perhaps more sensible to move this forward, since with python nothing works in the end.
And put it in `./CMakeLists.txt`

why?

> done as asked by the user.

please define what is asked by the user.

> So how about moving the `find_package(Python ...)` stuff up the tree?

i'd prefer `find_package()` where it's used.

Comment (wjwithagen) [reply cmake/modules/AddCephTest.cmake]: @tchaikov
I would move it up, to make sure that whenever there is a reference to a Python variable that Python detection has been done.  Just look in `./CMakeLists.txt` and see there are lots of packages detected (too) early. 
And like I said, you cannot build ceph without python
So I'm suggesting to move it up, to prevent you earlierer worries about referencing `Python_VERSION_*` without it being detected.

As for this actual PR, 
The user wants his stuff to build with the indicated python version.
In that sense it is silly/confusing to use something different for building some tools.
So perhaps build the env with 2.7 works, when 3.6 was asked.
But it is certainly confusing, and when there are errors, the kind of "oddities" only distract from getting to the real problem. So why not fix it?

Comment (tchaikov) [reply cmake/modules/AddCephTest.cmake]: > to make sure that whenever there is a reference to a Python variable that Python detection has been done. 

i think you need to be more specific. 

> Just look in `./CMakeLists.txt` and see there are lots of packages detected (too) early. 

i don't think we have to follow that it is. otherwise why are we improving Ceph?

> So I'm suggesting to move it up, to prevent you earlierer worries about referencing `Python_VERSION_*` without it being detected.

again, i don't want to search upwards/backwards to make sure that a variable is defined.

> The user wants his stuff to build with the indicated python version. In that sense it is silly/confusing to use something different for building some tools. So perhaps build the env with 2.7 works, when 3.6 was asked.

i think i've explain in https://github.com/ceph/ceph/pull/29761#discussion_r317050077. but probably you missed it?

> it is certainly confusing

please be specific what the confusion is.

Comment (wjwithagen) [reply cmake/modules/AddCephTest.cmake]: > > The user wants his stuff to build with the indicated python version. In that sense it is silly/confusing to use something different for building some tools. So perhaps build the env with 2.7 works, when 3.6 was asked.
> 
> i think i've explain in [#29761 (comment)](https://github.com/ceph/ceph/pull/29761#discussion_r317050077). but probably you missed it?

Ah, you want to use  `Python_EXECUTABLE`, that fell out of the scope of the window.
So it looked the same, but was not. That I can fix.

> > it is certainly confusing
> 
> please be specific what the confusion is.

I do not expect any references to python2.7 when I have said `WITH_PYTHON2=OFF` like:
```
Running virtualenv with interpreter /usr/local/bin/python2.7
New python executable in /home/jenkins/workspace/ceph-master36/build/mgr-ansible-virtualenv/bin/python2.7
Also creating executable in /home/jenkins/workspace/ceph-master36/build/mgr-ansible-virtualenv/bin/python
```
Or how about this one:
```
Requirement already satisfied: pathlib2>=2.2.0; python_version < "3.6" in /home/jenkins/workspace/ceph-master36/build/mgr-dashboard-virtualenv/lib/python2.7/site-packages (from pytest<4->-r requirements-test.txt (line 2)) (2.3.4)
```
the part with `python_version < "3.6"`  really makes me wonder.

Comment (wjwithagen) [reply cmake/modules/AddCephTest.cmake]: > > to make sure that whenever there is a reference to a Python variable that Python detection has been done.
> 
> i think you need to be more specific.
> 
> > Just look in `./CMakeLists.txt` and see there are lots of packages detected (too) early.
> 
> i don't think we have to follow that it is. otherwise why are we improving Ceph?
> 
> > So I'm suggesting to move it up, to prevent you earlierer worries about referencing `Python_VERSION_*` without it being detected.
> 
> again, i don't want to search upwards/backwards to make sure that a variable is defined.

What you suggested was putting a `find_package(Python....) ` in `cmake/modules/AddCephTest.cmake `??
That would obfuscate to point where the check for python is really made.

I do not really expect those kinds of checks/finds to be hidden in a module.
But if that is what you prefer, we can move (or add) it to `cmake/modules/AddCephTest.cmake `

Comment (wjwithagen) [reply cmake/modules/AddCephTest.cmake]: > > The user wants his stuff to build with the indicated python version. In that sense it is silly/confusing to use something different for building some tools. So perhaps build the env with 2.7 works, when 3.6 was asked.
> 
> i think i've explain in [#29761 (comment)](https://github.com/ceph/ceph/pull/29761#discussion_r317050077). but probably you missed it?

Ah, you want to use  `Python_EXECUTABLE`, that fell out of the scope of the window.
So it looked the same, but was not. That I can fix

> > it is certainly confusing
> 
> please be specific what the confusion is.

**Comment by epuertat** [comment]
Just a clarification: this was auto-closed by Github, as #29848 included the words `"Fixes #29761"`. I think, while unintentional, the outcome was the desired.

@wjwithagen: please, ensure now this passes in FreeBSD, or let us know otherwise.

File: cmake/modules/AddCephTest.cmake (hunk 1 above)
Comment (wjwithagen) [reply cmake/modules/AddCephTest.cmake]: @tchaikov
Is this now to your liking?
